Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactRandomCutTreeFloat.java

```diff
@@ -118,4 +119,43 @@ public CompactRandomCutTreeFloat build() {
             return new CompactRandomCutTreeFloat(this);
         }
     }
+
+    /**
+     * Return a new {@link Cut}, which is chosen uniformly over the space of
+     * possible cuts for the given bounding box. The cut is over single precision,
+     * although the representation is over doubles still.
+     *
+     * @param random A random number generator
+     * @param box    A bounding box that we want to find a random cut for.
+     * @return A new Cut corresponding to a random cut in the bounding box.
+     */
+    @Override
+    Cut treeCut(Random random, AbstractBoundingBox<?> box) {
+        double rangeSum = box.getRangeSum();
+        checkArgument(rangeSum > 0, "box.getRangeSum() must be greater than 0");
+
+        double breakPoint = random.nextDouble() * rangeSum;
+
+        for (int i = 0; i < box.getDimensions(); i++) {
+            double range = box.getRange(i);
+            if (breakPoint <= range) {
+                float cutValue = (float) (box.getMinValue(i) + breakPoint);
+
+                // Random cuts have to take a value in the half-open interval [minValue,
+                // maxValue) to ensure that a
+                // Node has a valid left child and right child.
+                if ((cutValue == box.getMaxValue(i)) && (box.getMinValue(i) < box.getMaxValue(i))) {
+                    cutValue = Math.nextAfter(cutValue, box.getMinValue(i));
```

Comment:
If min-values and max-values are both double-precision, then isn't it possible that taking a single-precision step will result in a cut value that's smaller than the min value?

Reply:
Yes it can. But minValues and maxValues are both float (as would be guaranteed for single precision) then there should be no problem. We can force an explicit conversion to make that point. However the bigger issue is that if max - min is a double and 0 as a float then the range calculations etc will all have problems.